lib/teleterm/gateway/gateway.go

```diff
@@ -116,29 +145,114 @@ func (g *Gateway) Close() error {
 
 // Serve starts the underlying ALPN proxy. Blocks until closeContext is canceled.
 func (g *Gateway) Serve() error {
-	g.Log.Info("Gateway is open.")
+	g.cfg.Log.Info("Gateway is open.")
 
 	if err := g.localProxy.Start(g.closeContext); err != nil {
 		return trace.Wrap(err)
 	}
 
-	g.Log.Info("Gateway has closed.")
+	g.cfg.Log.Info("Gateway has closed.")
+
+	return nil
+}
+
+func (g *Gateway) URI() uri.ResourceURI {
+	return g.cfg.URI
+}
+
+func (g *Gateway) SetURI(newURI uri.ResourceURI) {
+	g.cfg.URI = newURI
+}
+
+func (g *Gateway) TargetURI() string {
+	return g.cfg.TargetURI
+}
+
+func (g *Gateway) TargetName() string {
+	return g.cfg.TargetName
+}
+
+func (g *Gateway) Protocol() string {
+	return g.cfg.Protocol
+}
+
+func (g *Gateway) TargetUser() string {
+	return g.cfg.TargetUser
+}
+
+func (g *Gateway) TargetSubresourceName() string {
+	return g.cfg.TargetSubresourceName
+}
+
+func (g *Gateway) SetTargetSubresourceName(value string) {
+	g.cfg.TargetSubresourceName = value
+}
+
+func (g *Gateway) Log() *logrus.Entry {
+	return g.cfg.Log
+}
+
+func (g *Gateway) LocalAddress() string {
+	return g.cfg.LocalAddress
+}
+
+func (g *Gateway) LocalPort() string {
+	return g.cfg.LocalPort
+}
+
+// SetLocalPortAndRestart attempts to create a listener on the specified port. If successful, it
+// obtains a listener on the new port, stops the old proxy, updates the fields on gateway and then
+// starts the new proxy with the new listener. It starts the local proxy even if the previous one
+// wasn't started – SetLocalPortAndRestart is expected to be called after Serve.
+//
+// If it fails it's imperative that the current proxy is kept intact and the fields on gateway are
+// not changed. This way if the user attempts to change the port to one that cannot be obtained,
+// they're able to correct that mistake and choose a different port.
+//
+// SetLocalPortAndRestart is a noop if port is equal to the existing port.
+func (g *Gateway) SetLocalPortAndRestart(port string) error {
+	if port == g.cfg.LocalPort {
+		return nil
+	}
+
+	result, err := newListenerAndLocalProxy(*g.cfg, port)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
```

Comment:
> I don't think you need to call this, just closing the localProxy should be fine (this will be relevant if you use a single parent context, you won't be able to cancel it here)

Yeah, I noticed that it might be unnecessary too but so far I was just copying this around without thinking. I'll take care of that in the new version.